opamp/observiq/observiq_client.go

```diff
@@ -229,6 +238,91 @@ func (c *Client) onRemoteConfigHandler(ctx context.Context, remoteConfig *protob
+		Packages:                      map[string]*protobufs.PackageStatus{},
+	}
+
+	// Retrieve last known status (this should return with minimal info even on first time)
+	lastPackageStatuses, err := c.packagesStateProvider.LastReportedStatuses()
+
+	// If there is a problem retrieving the last saved PackageStatuses, we will log the error
+	// but continue on as the only thing missing will be the agent package hash.
+	if err != nil {
+		c.logger.Debug("Failed to retrieve last reported package statuses", zap.Error(err))
+	}
+
+	var lastPkgStatusMap map[string]*protobufs.PackageStatus
+	if lastPackageStatuses != nil {
+		lastPkgStatusMap = lastPackageStatuses.GetPackages()
+	}
+
+	curPackages := c.createPackageStatusMap(packagesAvailable.GetPackages(), lastPkgStatusMap)
+	curPackageStatuses.Packages = curPackages
+
+	if err = c.packagesStateProvider.SetLastReportedStatuses(curPackageStatuses); err != nil {
+		c.logger.Debug("Failed to save last reported package statuses", zap.Error(err))
+	}
+
+	if err = c.opampClient.SetPackageStatuses(curPackageStatuses); err != nil {
+		return fmt.Errorf("failed to set package statuses: %w", err)
+	}
+
+	return nil
+}
+
+func (c *Client) createPackageStatusMap(
+	pkgAvailMap map[string]*protobufs.PackageAvailable,
+	lastPkgStatusMap map[string]*protobufs.PackageStatus) map[string]*protobufs.PackageStatus {
+	pkgStatusMap := map[string]*protobufs.PackageStatus{}
+
+	// Loop through all of the available packages sent from the server
+	for name, availPkg := range pkgAvailMap {
+		switch name {
+		// If it's an expected package, return an installing status
+		case "observiq-otel-collector":
+			var agentHash []byte
+			if lastPkgStatusMap != nil && lastPkgStatusMap[name] != nil {
+				if lastPkgStatusMap[name].GetAgentHasVersion() != version.Version() {
+					c.logger.Debug(fmt.Sprintf(
+						"Version: %s and last reported package status version: %s differ",
+						version.Version(),
+						lastPkgStatusMap[name].GetAgentHasVersion()))
+				}
+				agentHash = lastPkgStatusMap[name].GetAgentHasHash()
+			}
+
+			pkgStatusMap[name] = &protobufs.PackageStatus{
+				Name:                 name,
+				AgentHasVersion:      version.Version(),
+				AgentHasHash:         agentHash,
+				ServerOfferedVersion: availPkg.GetVersion(),
+				ServerOfferedHash:    availPkg.GetHash(),
+				Status:               protobufs.PackageStatus_Installed,
+			}
+
+			if pkgStatusMap[name].GetAgentHasVersion() != pkgStatusMap[name].GetServerOfferedVersion() {
+				pkgStatusMap[name].Status = protobufs.PackageStatus_Installing
+			}
+		// If it's not an expected package, return a failed status
+		default:
+			pkgStatusMap[name] = &protobufs.PackageStatus{
+				Name:                 name,
+				ServerOfferedVersion: availPkg.GetVersion(),
+				ServerOfferedHash:    availPkg.GetHash(),
+				Status:               protobufs.PackageStatus_InstallFailed,
+				ErrorMessage:         fmt.Sprintf("Package %s not allowed", name),
```

Comment:
I would maybe change this message to `Package %s not supported`. `Allowed` make it seem like a permissions error.

Reply:
👍